### PR TITLE
Return408 for write due to client timed out

### PR DESCRIPTION
**Pull request by xinlian12**
For write operation with client side timed out, should return 408 instead of 410.

Matching code in .net: https://msdata.visualstudio.com/CosmosDB/_search?action=contents&text=TransportClient&type=code&lp=code-Project&filters=ProjectFilters%7BCosmosDB%7DRepositoryFilters%7BCosmosDB%7D&pageSize=25&result=DefaultCollection%2FCosmosDB%2FCosmosDB%2FGBmaster%2F%2FProduct%2FMicrosoft.Azure.Documents%2FSharedFiles%2FRntbd2%2FTransportClient.cs

### CODE REVIEW

**Review by moderakh — APPROVED** [review]
LGTM. thanks

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/directconnectivity/rntbd/RntbdRequestRecord.java [hunk 1]
```diff
@@ -226,9 +228,19 @@ public long transportRequestId() {
...
     public boolean expire() {
-        final GoneException error = new GoneException(this.toString(), null, this.args.physicalAddress());
-        BridgeInternal.setRequestHeaders(error, this.args.serviceRequest().getHeaders());
+        final CosmosException error;
+        if (this.args.serviceRequest().isReadOnly()) {
+            // Always converting RequestTimeoutException to GoneException for ReadOnly request. Since they can be safely retried.
+            error = new GoneException(this.toString(), null, this.args.physicalAddress());
+        } else if (!this.hasSendingRequestStarted()) {
+            // If request not sent for write request, converting to GoneException so it can be retried.
+            error = new GoneException(this.toString(), null, this.args.physicalAddress());
```
Comment (mbhaskar) [inline sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/directconnectivity/rntbd/RntbdRequestRecord.java:237]: GoneException in the cosmos context means that "The requested resource is no longer available at the server". Why are we reusing that exception here? I understand that GoneException is wired properly for retries but would'nt it make sense to have a client specific exception here which will make it easier to debug?
Thoughts?

Comment (mbhaskar) [inline sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/directconnectivity/rntbd/RntbdRequestRecord.java:237]: On a side note can we merge these two if conditions? (exclusive of above comment which is a bigger change)

Comment (xinlian12) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/directconnectivity/rntbd/RntbdRequestRecord.java:237]: yup, sure, will combine in the next iteration

Comment (xinlian12) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/directconnectivity/rntbd/RntbdRequestRecord.java:237]: @mbhaskar  I agree with u, we should have a better definition of the GoneException. Currently, we are kind a little bit abuse the use of it, basically we are using it to wire up the retries, address refresh etc. I think this should be a larger scope discussion about the design etc.

As for debugging, this one can be quickly identified through the  Rntbd diagnostics. But would a different substatus code help here?

Comment (mbhaskar) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/directconnectivity/rntbd/RntbdRequestRecord.java:237]: Yeah different substatus is definitely a good start to reduce the abuse. Eventually we can also introduce a new exception subclassing GoneException for the client generated scenarios

File: sdk/cosmos/azure-cosmos/src/test/java/com/azure/cosmos/implementation/directconnectivity/rntbd/RntbdRequestRecordTests.java [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+        RntbdRequestTimer requestTimer = new RntbdRequestTimer(5000, 5000);
+        RntbdRequestRecord record = new AsyncRntbdRequestRecord(requestArgs, requestTimer);
+        if (requestSent) {
+            record.setSendingRequestHasStarted();
+        }
+        record.expire();
+
+        try{
+            record.get();
+            Assert.fail("RntbdRequestRecord should complete with exception");
```
[the remark this answers is not in the document]
Comment (xinlian12) [reply sdk/cosmos/azure-cosmos/src/test/java/com/azure/cosmos/implementation/directconnectivity/rntbd/RntbdRequestRecordTests.java]: updated.

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/directconnectivity/rntbd/RntbdRequestRecord.java (hunk 1 above)
Comment (xinlian12) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/directconnectivity/rntbd/RntbdRequestRecord.java]: updated.

Comment (xinlian12) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/directconnectivity/rntbd/RntbdRequestRecord.java:236]: Discussed offline, agreed will leave this change out of this PR. 

Meanwhile, @mbhaskar created this working item to track it : https://github.com/Azure/azure-sdk-for-java/issues/20271

Thanks

**Pull request merged**